WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift

```diff
@@ -0,0 +1,237 @@
+            self?.knownReaderIDs = readerIDs
+            /// We are now ready to actually begin the search
+            self?.state = .beginSearch
+        }).store(in: &subscriptions)
+
+    }
+}
+
+private extension CardReaderConnectionController {
+    func didSetState() {
+        switch state {
+        case .beginSearch:
+            onBeginSearch()
+        case .searching:
+            onSearching()
+        case .endSearch:
+            onEndSearch()
+        case .foundReader:
+            onFoundReader()
+        case .connectToReader:
+            onConnectToReader()
+        case .failed(let error):
+            onFailed(error: error)
+        default:
+            dismissAnyModal()
+        }
+    }
+
+    /// Begin the search for a card reader
+    ///
+    func onBeginSearch() {
+        self.state = .searching
+        let action = CardPresentPaymentAction.startCardReaderDiscovery(
+            siteID: siteID,
+            onReaderDiscovered: { [weak self] cardReaders in
+                /// Surprisingly, onReaderDiscovered may be called with an empty array
+                ///
+                guard cardReaders.count > 0 else {
+                    return
+                }
+
+                /// For now, we only work with the first card reader returned
+                /// When we stop using proximity, we'll need more elaborate logic as we
+                /// won't be able to do that anymore
+                ///
+                self?.foundReader = cardReaders.first
+
+                guard let foundReaderID = self?.foundReader?.id else {
+                    DDLogWarn("onBeginSearch unexpectedly handled a nil found reader")
+                    self?.state = .endSearch
+                    return
+                }
+
+                /// If we know this reader, automatically connect to it
+                ///
+                let knownReaderIDs = self?.knownReaderIDs ?? []
+                if knownReaderIDs.contains(foundReaderID) {
+                    self?.state = .connectToReader
+                } else {
+                    self?.state = .foundReader
+                }
+            },
+            onError: { [weak self] error in
+                ServiceLocator.analytics.track(.cardReaderDiscoveryFailed, withError: error)
+                self?.state = .failed(error)
+            })
+
+        ServiceLocator.stores.dispatch(action)
+    }
+
+    /// A search is in progress
+    ///
+    func onSearching() {
+        alerts.scanningForReader(from: fromController, cancel: {
+            self.state = .endSearch
+        })
+    }
+
+    /// End the search for a card reader
+    ///
+    func onEndSearch() {
+        let action = CardPresentPaymentAction.cancelCardReaderDiscovery() {_ in
+            self.state = .idle
+        }
+        ServiceLocator.stores.dispatch(action)
+    }
+
+    /// A reader has been found
+    ///
+    func onFoundReader() {
+        guard foundReader != nil else {
```

Comment:
What about something like:
```
guard let foundReaderId = foundReader.id else {
}
```

That would allow packing the two guards into one

Reply:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/4721/commits/8d8e83d98fb675e5c59ae16a3420d6ce0e104236